### PR TITLE
Guarantee that a static IntegerExp is never changed

### PR DESCRIPTION
**Pull request by ghost**
This follows up https://github.com/dlang/dmd/pull/9500, as suggested by @JinShil.

~~A bool flag is set when creating a static IntegerExp.~~
~~It's used to make sure that neither the value nor the type are modified.~~
~~Any attempt will lead to a ICE.~~

~~Note a small drawback : before adding the flag the class instance size was exactly of 48 bytes.~~

### CODE REVIEW

File: src/dmd/expression.h [hunk 1]
```diff
@@ -231,6 +231,7 @@ class IntegerExp : public Expression
 {
 public:
     dinteger_t value;
+    bool literealFlag;
```
Comment (kinke) [inline src/dmd/expression.h]: Little typo. I think `isLiteral` would be clearer (although 'literal' itself might be considered confusing, as all `IntegerExp` are AST literals - `isReadonlyLiteral` might be another option).

I've just grepped for `setInteger`, and it looks like you replaced the only occurrence in `expressionsem.d`...

**Comment by WalterBright** [comment]
Isn't this doing what `const` is supposed to do?

**Comment by JinShil** [comment]
> Isn't this doing what `const` is supposed to do?

I believe so, but how does one properly express that with an `IntegerExp`?

File: src/dmd/astbase.d [hunk 2]
```diff
@@ -4383,7 +4384,13 @@ struct ASTBase
         {
             /* 'Normalize' the value of the integer to be in range of the type
              */
-            switch (type.toBasetype().ty)
+            const ty = type.toBasetype().ty;
+            if (literalFlag && ty != type.ty)
+            {
+                error("cannot modify an uncontextualized static IntegerExp");
```
Comment (thewilsonator) [inline src/dmd/astbase.d]: That is a horrible error message if someone actually manages to trip it. Maybe make it an ICE?

File: src/dmd/expression.h (hunk 1 above)
Comment (Basile-z) [reply src/dmd/expression.h]: Yes, almost a good reason to drop the flag if there was not the Expression.type, which is a public field

**Comment by Basile-z** [comment]
> > Isn't this doing what const is supposed to do?

> I believe so, but how does one properly express that with an IntegerExp?

It's not possible to return a const(IntegerExp) in IntegerExp.literal. This would require to change the parameters of several constructors, ( + the field of several classes), such as the ReturnStatement one.

**Comment by Basile-z** [comment]
Please, note that I've dropped the `bool` flag.

- The way `normalize()` is changed allows `setInteger()` to be private.
- I didn't like that more memory might be allocated (48 to 49 bytes meant 56 bytes actually allocated).
- public field `Expression.type` from the base class was not protected anyway.
- `normalize()` will never change the type `Tint32` set when the static instances are created. Even the ctor that just takes a value doesn't bother calling it (while the other ctor does).

I'm happy with the PR now so if you agree too with the change I'll just squash.

**Comment by WalterBright** [comment]
> how does one properly express that with an IntegerExp?

```
const dinteger_t value;
```

File: src/dmd/astbase.d [hunk 3]
```diff
@@ -4432,7 +4432,6 @@ struct ASTBase
                 else
                     assert(0);
                 break;
-
```
Comment (thewilsonator) [inline src/dmd/astbase.d]: ?

File: src/dmd/expression.d [hunk 4]
```diff
@@ -1818,70 +1819,69 @@ extern (C++) final class IntegerExp : Expression
...
     {
         /* 'Normalize' the value of the integer to be in range of the type
          */
-        dinteger_t result;
-        switch (ty)
+        switch (type.toBasetype().ty)
         {
         case Tbool:
-            result = (value != 0);
+            value = (value != 0);
```
Comment (JinShil) [inline src/dmd/expression.d]: This refactor seems largely unrelated.  Also please see #8 here: https://forum.dlang.org/post/p6oibo$1lmi$1@digitalmars.com

**Comment by Basile-z** [comment]
@thewilsonator and others, squashed and forced pushed, merge to your appreciation but don't forget that the stuff is already there, it just improve it, since #9500 was merged before I address the review comment from @JinShil. :wink:

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/expression.d [hunk 5]
```diff
@@ -1818,70 +1819,69 @@ extern (C++) final class IntegerExp : Expression
...
-    void setInteger(dinteger_t value)
+    private void setInteger(dinteger_t value)
     {
-        this.value = normalize(type.toBasetype().ty, value);
+        this.value = value;
+        normalize();
     }
 
-    static dinteger_t normalize(TY ty, dinteger_t value)
+    void normalize()
```
Comment (WalterBright) [inline src/dmd/expression.d]: This PR does not remove technical debt, it adds it. `normalize()` was a function that could be pure. It is refactored into a function that is not pure. I strongly object to this PR. Not only does it not do what the title says it does, I don't see a purpose to these changes, and none is offered.

Comment (WalterBright) [reply src/dmd/expression.d]: I see you did offer a rationale, making `setInteger()` private. I don't see a gain, though, as `normalize()` gets exposed instead. I don't see any improvement.

**Comment by WalterBright** [comment]
Closed because turning a pure function into an impure one is not where we should be going. Note that `normalize()` still can't be marked pure because `target.ptrsize` is a mutable global, even though it doesn't change after initialization. But at least the rest of it is pure.

**Comment by JinShil** [comment]
@WalterBright, please be aware that there is an issue with #9500.  That PR introduced static singleton instances of `IntegerExp` in an effort to (I think) reduce the memory footprint of DMD.  Those singleton instances are expected to have their `value` remain constant for their lifetime.  However, because `setInteger` is available, it invites the possibility of introducing a difficult to trace bug.

You might argue that #9500 should not have been merged.  @Basile-z even admitted that it was merged before he had time to address this issue.  This PR was an attempt to address the issue.  If this PR is not suitable, please suggest an alternative.

**Comment by kinke** [comment]
I'm also surprised this got dismissed that quickly (although I would have killed `setInteger()` altogether). What's a theoretically pure function worth if it's only used by the class itself (IIRC, just in the constructor and then in `toInteger()`), to normalize its (basically constant) value to the apparently mutable type (see comment in `toInteger()`), except for bloating the signature/call sites?

**Comment by Basile-z** [comment]
> in an effort to (I think) reduce the memory footprint of DMD

And compilation time

**Pull request closed without merging**